src/test/scala/is/hail/expr/ir/ExtractAggregatorsSuite.scala

```diff
@@ -0,0 +1,244 @@
+package is.hail.expr.ir
+
+import is.hail.asm4s._
+import is.hail.annotations._
+import ScalaToRegionValue._
+import is.hail.annotations.aggregators._
+import is.hail.check._
+import is.hail.expr.{TAggregable, TArray, TFloat64, TInt32, TStruct, RegionValueAggregator, hailType, HailRep, TBoolean, TFloat32, TInt64, Type}
+import is.hail.expr.ir._
+import org.testng.annotations.Test
+import org.scalatest._
+import Matchers._
+
+class ExtractAggregatorsSuite {
+
+  private def defaultValue(t: Type): Any = t match {
+    case _: TBoolean => false
+    case _: TInt32 => 0
+    case _: TInt64 => 0L
+    case _: TFloat32 => 0.0f
+    case _: TFloat64 => 0.0
+    case _ => 0L // reference types
+  }
+
+  private def packageResults(region: Region, t: TStruct, a: Array[RegionValueAggregator]): Long = {
+    val rvb = new RegionValueBuilder()
+    rvb.set(region)
+    rvb.start(t)
+    rvb.startStruct()
+    t.fields.map(_.typ).zip(a).foreach { case (t, agg) =>
+      rvb.addRegionValue(t, region, agg.result(region))
+    }
+    rvb.endStruct()
+    rvb.end()
+  }
+
+  private def loadIRIntermediate(typ: Type): (Region, Long) => Any = typ match {
+    case _: TBoolean =>
+      _.loadBoolean(_)
+    case _: TInt32 =>
+      _.loadInt(_)
+    case _: TInt64 =>
+      _.loadLong(_)
+    case _: TFloat32 =>
+      _.loadFloat(_)
+    case _: TFloat64 =>
+      _.loadDouble(_)
+    case _ =>
+      (_, off) => off
+  }
+
+  private def runStage1[T : HailRep : TypeInfo, U: HailRep : TypeInfo](region: Region, ir: IR, aOff: Long, scope1: Int => U): (IR, Long) = {
+    val tAgg = TAggregable(hailType[T], Map("scope0" -> (0, hailType[U])))
+
+    val aggFb = FunctionBuilder.functionBuilder[Region, Array[RegionValueAggregator], T, Boolean, U, Boolean, Unit]
+    Infer(ir, Some(tAgg))
+    val (post, aggResultStruct, aggregators) = ExtractAggregators(ir, tAgg, aggFb)
+    val seqOp = aggFb.result()()
+
+    val tArray = TArray(hailType[T])
+    val len = tArray.loadLength(region, aOff)
+    val loadT = loadIRIntermediate(hailType[T])
+    for (i <- 0 until len) {
+      val me = tArray.isElementMissing(region, aOff, i)
+      val e =
+        if (me) defaultValue(hailType[T]).asInstanceOf[T]
+        else loadT(region, tArray.loadElement(region, aOff, i)).asInstanceOf[T]
+      seqOp(region, aggregators,
+        e,
+        me,
+        scope1(i),
+        false)
+    }
+
+    val aggResultsOff = packageResults(region, aggResultStruct, aggregators)
+    (post, aggResultsOff)
+  }
+
+  private def compileStage0[R: TypeInfo](ir: IR): AsmFunction3[Region, Long, Boolean, R] = {
+    val fb = FunctionBuilder.functionBuilder[Region, Long, Boolean, R]
+    // nb: inference is done by stage1
+    Compile(ir, fb)
+    fb.result()()
+  }
+
+  private def run[T : HailRep : TypeInfo, U : HailRep : TypeInfo, R : TypeInfo](region: Region, ir: IR, aOff: Long, scope1: Int => U): R = {
```

Comment:
Runs a possibly aggregating `ir` using the array at `aOff` in `region` as the aggregable. There's one name in scope called "scope0" and its value is determined by `scope0: Int => U`.